File: src/Configuration/KnownSettings.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

/// <summary>
/// Defines known configuration settings with standardized formats and access rules.
/// </summary>
public static class KnownSettings
{
    #region Setting Keys (Dot Notation)

    // Azure OpenAI settings
    public const string AzureOpenAIApiKey = "Azure.OpenAI.ApiKey";
    public const string AzureOpenAIEndpoint = "Azure.OpenAI.Endpoint";
    public const string AzureOpenAIChatDeployment = "Azure.OpenAI.ChatDeployment";
    
    // OpenAI settings
    public const string OpenAIApiKey = "OpenAI.ApiKey";
    public const string OpenAIChatModelName = "OpenAI.ChatModelName";
    
    // GitHub settings
    public const string GitHubToken = "GitHub.Token";
    
    // Copilot settings
    public const string CopilotModelName = "Copilot.ModelName";
    public const string CopilotApiEndpoint = "Copilot.ApiEndpoint";
    public const string CopilotIntegrationId = "Copilot.IntegrationId";
    public const string CopilotHmacKey = "Copilot.HmacKey";
    public const string CopilotEditorVersion = "Copilot.EditorVersion";

    // Application settings
    public const string AppMaxTokens = "App.MaxTokens";
    public const string AppPreferredProvider = "App.PreferredProvider";
    public const string AppAutoSaveChatHistory = "App.AutoSaveChatHistory";
    public const string AppAutoSaveTrajectory = "App.AutoSaveTrajectory";
    public const string AppChatCompletionTimeout = "App.ChatCompletionTimeout";
    
    #endregion
    
    #region Secret Settings
    
    /// <summary>
    /// Collection of known setting names that should be treated as secrets.
    /// </summary>
    private static readonly HashSet<string> _secretSettings = new(StringComparer.OrdinalIgnoreCase)
    {
        // Azure OpenAI secrets
        AzureOpenAIApiKey,
        
        // OpenAI secrets
        OpenAIApiKey,
        
        // GitHub secrets
        GitHubToken,
        
        // Copilot secrets
        CopilotHmacKey
    };
    
    #endregion
    
    #region Setting Format Mappings
    
    /// <summary>
    /// Mapping from dot notation to environment variable format.
    /// </summary>
    private static readonly Dictionary<string, string> _dotToEnvVarMap = new(StringComparer.OrdinalIgnoreCase)
    {
        // Azure OpenAI mappings
        { AzureOpenAIApiKey, "AZURE_OPENAI_API_KEY" },
        { AzureOpenAIEndpoint, "AZURE_OPENAI_ENDPOINT" },
        { AzureOpenAIChatDeployment, "AZURE_OPENAI_CHAT_DEPLOYMENT" },
        
        // OpenAI mappings
        { OpenAIApiKey, "OPENAI_API_KEY" },
        { OpenAIChatModelName, "OPENAI_CHAT_MODEL_NAME" },
        
        // GitHub mappings
        { GitHubToken, "GITHUB_TOKEN" },
        
        // Copilot mappings
        { CopilotModelName, "COPILOT_MODEL_NAME" },
        { CopilotApiEndpoint, "COPILOT_API_ENDPOINT" },
        { CopilotIntegrationId, "COPILOT_INTEGRATION_ID" },
        { CopilotHmacKey, "COPILOT_HMAC_KEY" },
        { CopilotEditorVersion, "COPILOT_EDITOR_VERSION" },
        
        // Application settings
        { AppMaxTokens, "APP_MAX_TOKENS" },
        { AppPreferredProvider, "CHATX_PREFERRED_PROVIDER" },
        { AppAutoSaveChatHistory, "CHATX_AUTO_SAVE_CHAT_HISTORY" },
        { AppAutoSaveTrajectory, "CHATX_AUTO_SAVE_TRAJECTORY" },
        { AppChatCompletionTimeout, "CHATX_CHAT_COMPLETION_TIMEOUT" }
    };
    
    /// <summary>
    /// Mapping from dot notation to CLI option format.
    /// </summary>
    private static readonly Dictionary<string, string> _dotToCLIOptionMap = new(StringComparer.OrdinalIgnoreCase)
    {
        // Azure OpenAI mappings
        { AzureOpenAIApiKey, "--azure-openai-api-key" },
        { AzureOpenAIEndpoint, "--azure-openai-endpoint" },
        { AzureOpenAIChatDeployment, "--azure-openai-chat-deployment" },
        
        // OpenAI mappings
        { OpenAIApiKey, "--openai-api-key" },
        { OpenAIChatModelName, "--openai-chat-model-name" },
        
        // GitHub mappings
        { GitHubToken, "--github-token" },
        
        // Copilot mappings
        { CopilotModelName, "--copilot-model-name" },
        { CopilotApiEndpoint, "--copilot-api-endpoint" },
        { CopilotIntegrationId, "--copilot-integration-id" },
        { CopilotHmacKey, "--copilot-hmac-key" },
        { CopilotEditorVersion, "--copilot-editor-version" },
        
        // Application settings
        { AppMaxTokens, "--max-tokens" },
        { AppAutoSaveChatHistory, "--auto-save-chat-history" },
        { AppAutoSaveTrajectory, "--auto-save-trajectory" },
        { AppChatCompletionTimeout, "--chat-completion-timeout" }
    };
    
    /// <summary>
    /// Mapping from environment variable format to dot notation.
    /// </summary>
    private static readonly Dictionary<string, string> _envVarToDotMap;
    
    /// <summary>
    /// Mapping from CLI option format to dot notation.
    /// </summary>
    private static readonly Dictionary<string, string> _cliOptionToDotMap;
    
    #endregion
    
    #region Category Groupings
    
    /// <summary>
    /// Collection of settings for Azure OpenAI integration.
    /// </summary>
    public static readonly HashSet<string> AzureOpenAISettings = new(StringComparer.OrdinalIgnoreCase)
    {
        AzureOpenAIApiKey,
        AzureOpenAIEndpoint,
        AzureOpenAIChatDeployment
    };
    
    /// <summary>
    /// Collection of settings for OpenAI integration.
    /// </summary>
    public static readonly HashSet<string> OpenAISettings = new(StringComparer.OrdinalIgnoreCase)
    {
        OpenAIApiKey,
        OpenAIChatModelName
    };
    
    /// <summary>
    /// Collection of settings for GitHub integration.
    /// </summary>
    public static readonly HashSet<string> GitHubSettings = new(StringComparer.OrdinalIgnoreCase)
    {
        GitHubToken
    };
    
    /// <summary>
    /// Collection of settings for Copilot integration.
    /// </summary>
    public static readonly HashSet<string> CopilotSettings = new(StringComparer.OrdinalIgnoreCase)
    {
        CopilotModelName,
        CopilotApiEndpoint,
        CopilotIntegrationId,
        CopilotHmacKey,
        CopilotEditorVersion
    };
    
    /// <summary>
    /// Collection of settings for application configuration.
    /// </summary>
    public static readonly HashSet<string> AppSettings = new(StringComparer.OrdinalIgnoreCase)
    {
        AppMaxTokens,
        AppPreferredProvider,
        AppAutoSaveChatHistory,
        AppAutoSaveTrajectory,
        AppChatCompletionTimeout
    };
    
    #endregion
    
    static KnownSettings()
    {
        // Initialize reverse mappings
        _envVarToDotMap = _dotToEnvVarMap.ToDictionary(kvp => kvp.Value, kvp => kvp.Key, StringComparer.OrdinalIgnoreCase);
        _cliOptionToDotMap = _dotToCLIOptionMap.ToDictionary(kvp => kvp.Value, kvp => kvp.Key, StringComparer.OrdinalIgnoreCase);
    }
    
    /// <summary>
    /// Checks if the given key is a known setting.
    /// </summary>
    /// <param name="key">The key to check (in any format).</param>
    /// <returns>True if the key is a known setting, false otherwise.</returns>
    public static bool IsKnown(string key)
    {
        if (_dotToEnvVarMap.ContainsKey(key)) return true;
        if (_envVarToDotMap.ContainsKey(key)) return true;
        if (_cliOptionToDotMap.ContainsKey(key)) return true;
        return false;
    }
    
    /// <summary>
    /// Checks if the given key is a secret value that should be obfuscated.
    /// </summary>
    /// <param name="key">The key to check (in any format).</param>
    /// <returns>True if the key is a secret, false otherwise.</returns>
    public static bool IsSecret(string key)
    {
        var dotNotationKey = ToDotNotation(key);
        return _secretSettings.Contains(dotNotationKey);
    }
    
    /// <summary>
    /// Gets all known setting names in dot notation format.
    /// </summary>
    /// <returns>Collection of all known setting names.</returns>
    public static IEnumerable<string> GetAllKnownSettings()
    {
        return _dotToEnvVarMap.Keys;
    }
    
    /// <summary>
    /// Converts a setting name to environment variable format.
    /// </summary>
    /// <param name="key">The setting name in any format.</param>
    /// <returns>The equivalent environment variable name.</returns>
    public static string ToEnvironmentVariable(string key)
    {
        if (IsEnvironmentVariableFormat(key)) return key;
        
        // Try to map to an environment variable
        var dotNotationKey = ToDotNotation(key);
        if (_dotToEnvVarMap.TryGetValue(dotNotationKey, out string? envVarKey))
        {
            return envVarKey;
        }

        // Otherwise, just return it
        return key;
    }
    
    /// <summary>
    /// Converts a setting name to CLI option format.
    /// </summary>
    /// <param name="key">The setting name in any format.</param>
    /// <returns>The equivalent CLI option.</returns>
    public static string ToCLIOption(string key)
    {
        if (IsCLIOptionFormat(key)) return key;
        
        // Try to map to a CLI option
        var dotNotationKey = ToDotNotation(key);
        if (_dotToCLIOptionMap.TryGetValue(dotNotationKey, out string? cliOption))
        {
            return cliOption;
        }
        
        // Otherwise, use a general algorithm
        var parts = dotNotationKey.Split('.');
        var kebabParts = parts.Select(p => ToKebabCase(p));
        return "--" + string.Join("-", kebabParts).ToLowerInvariant();
    }
    
    /// <summary>
    /// Converts a setting name to dot notation format.
    /// </summary>
    /// <param name="key">The setting name in any format.</param>
    /// <returns>The equivalent dot notation.</returns>
    public static string ToDotNotation(string key)
    {
        // If it's an environment variable format
        if (IsEnvironmentVariableFormat(key))
        {
            if (_envVarToDotMap.TryGetValue(key, out string? dotNotation))
            {
                return dotNotation;
            }
        }
        
        // If it's a CLI option format
        if (IsCLIOptionFormat(key))
        {
            if (_cliOptionToDotMap.TryGetValue(key, out string? dotNotation))
            {
                return dotNotation;
            }
            
            // Remove leading -- and convert kebab-case to PascalCase with dots
            var trimmed = key.TrimStart('-');
            var parts = trimmed.Split('-');
            for (int i = 0; i < parts.Length; i++)
            {
                if (parts[i].Length > 0)
                {
                    parts[i] = char.ToUpper(parts[i][0]) + parts[i].Substring(1).ToLower();
                }
            }
            
            return string.Join(".", parts);
        }

        return key;
    }
    
    /// <summary>
    /// Determines if the given key is in environment variable format.
    /// </summary>
    /// <param name="key">The key to check.</param>
    /// <returns>True if the key is in environment variable format, false otherwise.</returns>
    public static bool IsEnvironmentVariableFormat(string key)
    {
        return Regex.IsMatch(key, "^[A-Z0-9_]+$");
    }
    
    /// <summary>
    /// Determines if the given key is in CLI option format.
    /// </summary>
    /// <param name="key">The key to check.</param>
    /// <returns>True if the key is in CLI option format, false otherwise.</returns>
    public static bool IsCLIOptionFormat(string key)
    {
        return key.StartsWith("--");
    }
    
    /// <summary>
    /// Converts a string from PascalCase to kebab-case.
    /// </summary>
    /// <param name="input">The PascalCase string to convert.</param>
    /// <returns>The kebab-case version of the input string.</returns>
    private static string ToKebabCase(string input)
    {
        if (string.IsNullOrEmpty(input))
            return input;
        
        // Insert a hyphen before each uppercase letter that follows a lowercase letter
        var result = Regex.Replace(input, "(?<!^)([A-Z])", "-$1");
        return result.ToLowerInvariant();
    }
}

File: ts/src/Configuration/KnownSettings.ts
/**
 * Known configuration settings and their properties.
 */
export class KnownSettings {
  private static readonly _knownSettings: Map<string, SettingInfo> = new Map([
    // Provider settings
    ['provider', { canonical: 'provider', isSecret: false, description: 'AI provider to use' }],
    
    // GitHub Copilot settings
    ['github.copilot.token', { canonical: 'github.copilot.token', isSecret: true, description: 'GitHub Copilot token' }],
    ['github.copilot.endpoint', { canonical: 'github.copilot.endpoint', isSecret: false, description: 'GitHub Copilot endpoint' }],
    
    // OpenAI settings  
    ['OpenAI.ApiKey', { canonical: 'OpenAI.ApiKey', isSecret: true, description: 'OpenAI API key' }],
    ['OpenAI.Endpoint', { canonical: 'OpenAI.Endpoint', isSecret: false, description: 'OpenAI API endpoint' }],
    ['OpenAI.ChatModelName', { canonical: 'OpenAI.ChatModelName', isSecret: false, description: 'OpenAI chat model name' }],
    
    // Azure OpenAI settings
    ['azure.openai.token', { canonical: 'azure.openai.token', isSecret: true, description: 'Azure OpenAI token' }],
    ['azure.openai.endpoint', { canonical: 'azure.openai.endpoint', isSecret: false, description: 'Azure OpenAI endpoint' }],
    ['azure.openai.deployment', { canonical: 'azure.openai.deployment', isSecret: false, description: 'Azure OpenAI deployment name' }],
    
    // Chat history settings
    ['chat.history.directory', { canonical: 'chat.history.directory', isSecret: false, description: 'Directory for chat history files' }],
    ['chat.history.max.files', { canonical: 'chat.history.max.files', isSecret: false, description: 'Maximum number of chat history files to keep' }],
    
    // System settings
    ['system.prompt.default', { canonical: 'system.prompt.default', isSecret: false, description: 'Default system prompt' }],
    ['system.debug', { canonical: 'system.debug', isSecret: false, description: 'Enable debug output' }],
    ['system.quiet', { canonical: 'system.quiet', isSecret: false, description: 'Suppress verbose output' }],
    
    // Timeout settings
    ['timeout.request', { canonical: 'timeout.request', isSecret: false, description: 'Request timeout in seconds' }],
    ['timeout.response', { canonical: 'timeout.response', isSecret: false, description: 'Response timeout in seconds' }],
  ]);

  private static readonly _aliases: Map<string, string> = new Map([
    // Provider aliases
    ['ai.provider', 'provider'],
    ['ai_provider', 'provider'],
    
    // GitHub aliases
    ['github_copilot_token', 'github.copilot.token'],
    ['copilot.token', 'github.copilot.token'],
    ['copilot_token', 'github.copilot.token'],
    
    // OpenAI aliases
    ['openai_token', 'openai.token'],
    ['openai_api_key', 'openai.token'],
    ['openai.api.key', 'openai.token'],
    ['openai_endpoint', 'openai.endpoint'],
    ['openai_model', 'openai.model'],
    
    // Azure aliases
    ['azure_openai_token', 'azure.openai.token'],
    ['azure_openai_key', 'azure.openai.token'],
    ['azure.openai.key', 'azure.openai.token'],
    ['azure_openai_endpoint', 'azure.openai.endpoint'],
    ['azure_openai_deployment', 'azure.openai.deployment'],
    
    // System aliases
    ['debug', 'system.debug'],
    ['quiet', 'system.quiet'],
  ]);

  /**
   * Checks if a setting is a known configuration setting.
   */
  static isKnown(key: string): boolean {
    const normalized = this.toDotNotation(key);
    // Check with case-insensitive comparison
    for (const [knownKey] of this._knownSettings) {
      if (knownKey.toLowerCase() === normalized.toLowerCase()) return true;
    }
    for (const [aliasKey] of this._aliases) {
      if (aliasKey.toLowerCase() === normalized.toLowerCase()) return true;
    }
    return false;
  }

  /**
   * Gets the canonical form of a setting key.
   */
  static getCanonicalForm(key: string): string {
    const normalized = this.toDotNotation(key);
    
    // Check aliases first with case-insensitive comparison
    for (const [aliasKey, canonicalKey] of this._aliases) {
      if (aliasKey.toLowerCase() === normalized.toLowerCase()) {
        return canonicalKey;
      }
    }
    
    // Check known settings with case-insensitive comparison  
    for (const [knownKey] of this._knownSettings) {
      if (knownKey.toLowerCase() === normalized.toLowerCase()) {
        return knownKey; // Return the original casing from known settings
      }
    }
    
    return key; // Return original if not known
  }

  /**
   * Checks if a setting is a secret (should be masked in display).
   */
  static isSecret(key: string): boolean {
    const canonical = this.getCanonicalForm(key);
    const settingInfo = this._knownSettings.get(canonical);
    return settingInfo?.isSecret ?? false;
  }

  /**
   * Gets the description for a setting.
   */
  static getDescription(key: string): string | undefined {
    const canonical = this.getCanonicalForm(key);
    const settingInfo = this._knownSettings.get(canonical);
    return settingInfo?.description;
  }

  /**
   * Gets all known setting keys.
   */
  static getAllKnownSettings(): string[] {
    return Array.from(this._knownSettings.keys()).sort();
  }

  /**
   * Converts a key to dot notation format.
   */
  static toDotNotation(key: string): string {
    // Keep the original casing but convert underscores to dots
    return key.replace(/_/g, '.');
  }

  /**
   * Converts a key from dot notation to underscore format.
   */
  static toUnderscoreNotation(key: string): string {
    return key.toLowerCase().replace(/\./g, '_');
  }
}

interface SettingInfo {
  canonical: string;
  isSecret: boolean;
  description: string;
}